### ISSUE TITLE
convolve API feature request - cupyx.scipy.ndimage.convolve modes valid, full, same

### ISSUE THREAD

**Issue by randompast** [cupy/cupy#3414]
- CuPy version = 7.5.0
  - OS/Platform = Ubuntu 18.0.4
  - CUDA version = 10.2.89

Request adding modes: mode='valid', mode='full', and mode='same'.  This would match with numpy.convolve and scipy.convolve.

**Comment by kmaehashi** [comment]
`cupy.convolve` is work in progress: https://github.com/cupy/cupy/pull/3371

**Comment by randompast** [comment]
Understood, thanks @kmaehashi 

A temporary hack for mode='valid':
cupyx.scipy.ndimage.convolve(x, k)[ k.size/2 : -k.size/2 ]

Currently mode='constant', origin=0 or -1 matches mode='same'

Since the output is the length of the signal x, I'm not sure how to get full if I wanted it.

There's another potential to do convolutions using stride_tricks.as_strided and einsum.  I've noticed it runs pretty fast in comparison to many other convolution approaches.

**Comment by coderforlife** [comment]
If you wanted to implement this as a drop-in replacement, I made this function for use as `cupy.convolve`:

```python
def convolve(a, v, mode='full'):
    if mode not in ('full', 'same', 'valid'): raise ValueError('mode') 

    # Check both arrays and make sure `a` is the larger array
    if a.ndim != 1: raise ValueError('a must be 1D')
    if v.ndim != 1: raise ValueError('v must be 1D')
    if len(a) == 0: raise ValueError('a cannot be empty')
    if len(v) == 0: raise ValueError('v cannot be empty')
    if len(v) > len(a): a, v = v, a

    # Deal with full padding of data
    if mode == 'full':
        pad = len(v) - 1
        a = cupy.pad(a, ((pad+1)//2, pad//2), 'constant')
    
    # Perform convolution
    out = cupyx.scipy.ndimage.convolve1d(a, v, mode='constant')

    # Return result
    if mode == 'valid':
        remove = len(v) - 1
        out = out[remove//2 : -(remove+1)//2]
    return out
```

There may be a few off-by-1-errors in the above code (for example, it seems that odd-sized kernels may need an origin of 1 according to @randompast)

(note: as written this uses the soon-to-be-published `cupyx.scipy.ndimage.convolve1d`, you can temporarily use the `cupyx.scipy.ndimage.convolve` instead)

**Comment by randompast** [comment]
Nice job with that, I like your solution @coderforlife !  I expanded on what I was doing and compared it to other implementations.  I was only interested in the valid mode for 1D (1st, 2nd, 3rd) order and 3d 1st order.  Here's the link to that repo:
https://github.com/randompast/python-convolution-comparisons

**Comment by randompast** [comment]
The issue brought up with the original comment wasn't particularly clear with the 1d case.  It gets increasingly awkward as the dimensions increase.  The 3d version was the following, you can imagine how I might want to use cupyx.scipy.ndimage.convolve(x,k,'valid') instead of teasing apart this middle region.

```
def test_3d1o_cpv(x, k):
    return cupyx.scipy.ndimage.convolve(x, k)[ k.shape[0]/2 + k.shape[0]%2 -1 : -k.shape[0]/2
        , k.shape[1]/2 + k.shape[1]%2 -1
        , k.shape[2]/2 + k.shape[2]%2 -1 ]
```

**Comment by kmaehashi** [comment]
Closing as implemented in #3371
